samples/bluetooth/hci_uart/src/main.c

```diff
@@ -49,78 +55,136 @@ static K_FIFO_DEFINE(tx_queue);
  */
 #define H4_DISCARD_LEN 33
 
-static int h4_read(struct device *uart, uint8_t *buf,
-		   size_t len, size_t min)
-{
-	int total = 0;
-
-	while (len) {
-		int rx;
+/* Union used for variable into which packet header is received. */
+union bt_hci_hdr {
+	struct bt_hci_cmd_hdr cmd;
+	struct bt_hci_acl_hdr acl;
+	uint8_t raw[0];
```

Comment:
Do you really need to define an union for this (with a zero length array which is a gcc extension IIRC)?
Could not `hdr_buf[MAX(sizeof(struct bt_hci_cmd_hdr), sizeof(struct bt_hci_cmd_acl)]` be used instead?

Reply:
I agree, and then cast it where required. Also, you should use `[]` instead of `[0]` if you are gonna use that.

Reply:
did as you suggested, i need to cast it to structure to get payload length but that is only in one place (`get_len()`) and I used casting there.